[core] Prepend "use-client" directive + add docs and examples for using MUI libraries with Next.js App Router (#37656)
mui/material-ui: docs/data/material/guides/next-js-app-router/next-js-app-router.md

> # Next.js App Router

Author: I agree it's a better user experience to have a different doc per library, the only content that will be duplicated could be the Joy UI setup part and the introduction

<p class="description">Learn how to use MUI libraries with the Next.js App Router</p>

## Next.js and React Server Components

> Next.js's App Router implements React Server Components, which is a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) in React.

Reviewer: ```suggestion
The Next.js App Router implements React Server Components, a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) introduced in React 18.
```

> To support the App Router, currently all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are designated as client components and exported with the "use client" directive.

Reviewer: ```suggestion
To support the App Router, all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are designated as Client Components and exported with the `"use client"` directive.
```

:::info

> Don't need a guide? Use one of these example repos and get started:

Reviewer: Rather than framing it in terms of need/skill, how about something like this?

```suggestion
Starting fresh on a new App Router-based project?
Jump right into the code with one of these example repos:
```

- [Material UI](https://github.com/mui/material-ui/blob/master/examples/material-next-app-router-ts)
- [Joy UI](https://github.com/mui/material-ui/blob/master/examples/joy-next-app-router-ts)

> - [Base UI with Tailwind CSS](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts)

Reviewer 2: Would be great to add additional info about the product because some developers might have seen these names the first time.

e.g.
```md
- [Joy UI](…): Our new in-house library with fresh design.
- [Base UI with Tailwind CSS]: A collection of unstyled components.
```

cc @samuelsycamore

Author: I've updated it to:
```
Material UI - A React UI library that implements Material Design
Joy UI - Our newest customizable UI library designed to spark joy
Base UI – A collection of headless components and hooks
```

  :::

## Using Material UI with the default theme

> If you are using the default theme, Material UI components will work out-of-the-box in Next.js routing files such as `layout.js` or `page.js` - which are server components by default.

Reviewer: Active voice.

```suggestion
If you're using the default theme, you can add Material UI components to Next.js routing files such as `layout.js` or `page.js` (which are Server Components by default) without any additional configuration, as shown below:
```

<!-- TODO: investigate whether it still needs an explicit <head/> to prevent FOUC https://github.com/mui/material-ui/issues/34905#issuecomment-1332040656 -->

```jsx
// app/layout.js - no directives needed
export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}

// app/page.js - no directives needed
import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import Container from "@mui/material/Container";
import Typography from "@mui/material/Typography";

export default function Home() {
  return (
    <main>
      <Container>
        <Box>
          <Card>
            <Typography variant="h2">Hello World ~</Typography>
          </Card>
        </Box>
      </Container>
    </main>
  );
}
```

## Using Material UI with a custom theme

### Theme Registry

> To set up the theme context, create a `ThemeRegistry` component that combines the Emotion CacheProvider, the Material UI Theme Provider and the `useServerInsertedHTML` hook from `next/navigation` as follows:

Reviewer: ```suggestion
To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider`, and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```

    cache.insert = (...args) => {
      const serialized = args[1];
      if (cache.inserted[serialized.name] === undefined) {
        inserted.push(serialized.name);
      }
      return prevInsert(...args);
    };
    const flush = () => {
      const prevInserted = inserted;
      inserted = [];
      return prevInserted;
    };
    return { cache, flush };
  });

  useServerInsertedHTML(() => {
    const names = flush();
    if (names.length === 0) {
      return null;
    }
    let styles = '';
    for (const name of names) {
      styles += cache.inserted[name];
    }
    return (
      <style
        key={cache.key}
        data-emotion={`${cache.key} ${names.join(' ')}`}
        dangerouslySetInnerHTML={{
          __html: styles,
        }}
      />
    );
  });

  return (
    <CacheProvider value={cache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
}

// app/layout.js
export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>
        <ThemeRegistry options={{ key: 'mui' }}>{children}</ThemeRegistry>
      </body>
    </html>
  );
}
```

> ### CSS Injection Order

Reviewer: Sentence case (rather than Title Case) for headlines.

```suggestion
### CSS injection order
```

<!-- https://github.com/emotion-js/emotion/issues/3059 -->

> By default emotion will inject Material UI styles at the bottom of the HTML `<head>`, which gives it precedence over custom styles (e.g. from CSS modules, Tailwind CSS, or even plain CSS).

Reviewer: ```suggestion
By default, Emotion injects Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles (for example, from CSS modules, Tailwind CSS, or plain CSS).
```

> In regular client-side React, the `prepend: true` option is passed to `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.

Reviewer: It's not "regular" anymore! 🤪🙃

```suggestion
In client-side React, the `prepend: true` option is passed to `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.
```

> Currently `prepend` does not work reliably with the App Router, but you can work around it by wrapping emotion styles in a CSS `@layer` with a modification to the snippet above:

Reviewer: ```suggestion
Currently, `prepend` does not work reliably with the App Router, but you can work around it by wrapping Emotion styles in a CSS `@layer` with a modification to the snippet above:
```

```diff
 useServerInsertedHTML(() => {
   const names = flush();
   if (names.length === 0) {
     return null;
   }
   let styles = '';
   for (const name of names) {
     styles += cache.inserted[name];
   }
   return (
     <style
       key={cache.key}
       data-emotion={`${cache.key} ${names.join(' ')}`}
       dangerouslySetInnerHTML={{
-        __html: styles,
+        __html: options.prepend ? `@layer emotion {${styles}}` : styles,
       }}
     />
   );
 });
```

## Using Joy UI

> The set up for Joy UI is the same as [Material UI with a custom theme](#using-material-ui-with-a-custom-theme), except replace Material's `ThemeProvider` with Joy UI's `CssVarsProvider`:

Reviewer: ```suggestion
The setup for Joy UI is the same as [Material UI with a custom theme](#using-material-ui-with-a-custom-theme), except you must replace Material UI's `ThemeProvider` with Joy UI's `CssVarsProvider`:
```

```diff
 // ThemeRegistry
-import { ThemeProvider } from '@mui/material/styles';
+import { CssVarsProvider } from '@mui/joy/styles';
```

See this [example repo](https://github.com/mui/material-ui/blob/master/examples/joy-next-app-router-ts) for a working demo of Joy UI with the Next.js App Router.

## Using Base UI

> Base UI is styling-solution-agnostic, if you are using Emotion you can follow the same steps above.

Reviewer: ```suggestion
Base UI is style-agnostic; if you're using Emotion, then you can follow the same steps outlined for Material UI.
```

> If you are using Tailwind CSS, see this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a working demo.

Reviewer: ```suggestion
If you're using Tailwind CSS, see this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a working demo.
```

### Customizing Base UI with function props

> A common customization method in Base UI is to pass a function to slots in `slotProps` in order to apply dynamic props to slots. For example, changing the background color by applying a different class when a Button is disabled:

Reviewer: To make this a complete sentence.

```suggestion
A common customization method in Base UI is to pass a function to slots in `slotProps` in order to apply dynamic props to slots. For example, you might want to change the background color by applying a different class when a Button is disabled:
```

```tsx
// page.tsx

export default function Page() {
  return (
    <>
      {/* Next.js won't render this button without 'use-client'*/}
      <Button
        slotProps={{
          root: (ownerState: ButtonOwnerState) => ({
            className: ownerState.disabled ? 'bg-gray-400' : 'bg-blue-400',
          }),
        }}
      >
        Submit
      </Button>

      {/* Next.js can render this */}
      <Button
        slotProps={{
          root: {
            className: 'bg-gray-400',
          },
        }}
      >
        Return
      </Button>
    </>
  );
}
```

> However, this will not work since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).

Reviewer: ```suggestion
Unfortunately, **this will not work** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

> It is recommended that these components be ["moved to the leaves"](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves) to avoid this issue and improve overall performance.

Reviewer: Who recommends this?

```suggestion
Instead, the Next.js team recommends moving components like these ["to the leaves"](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves) to avoid this issue and improve overall performance.
```

# Next.js App Router

<p class="description">Learn how to use Material UI with the Next.js App Router.</p>

:::info
Starting fresh on a new App Router-based project?
Jump right into the code with one of these example repos:

- [Material UI - A React UI library that implements Material Design](https://github.com/mui/material-ui/blob/master/examples/material-next-app-router-ts)
- [Joy UI - Our newest customizable UI library designed to spark joy](https://github.com/mui/material-ui/blob/master/examples/joy-next-app-router-ts)

> - [Base UI – A collection of headless components and hooks](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts)

Reviewer: We can drop these links here, since the doc is just about Material UI now.

  :::

## Next.js and React Server Components

The Next.js App Router implements React Server Components, a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) introduced in React 18.

To support the App Router, currently all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are exported with the `"use client"` directive.

> :::warning

Reviewer: same as Base comments

:::

## Using Material UI with the default theme

If you're using the default theme, you can add Material UI components to Next.js routing files such as `layout.js` or `page.js` (which are Server Components by default) without any additional configuration, as shown below:

<!-- TODO: investigate whether it still needs an explicit <head/> to prevent FOUC https://github.com/mui/material-ui/issues/34905#issuecomment-1332040656 -->

```jsx
// app/layout.js - no directives needed
export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}

// app/page.js - no directives needed
import Box from "@mui/material/Box";
import Card from "@mui/material/Card";
import Container from "@mui/material/Container";
import Typography from "@mui/material/Typography";

export default function Home() {
  return (
    <main>
      <Container>
        <Box>
          <Card>
            <Typography variant="h2">Hello World ~</Typography>
          </Card>
        </Box>
      </Container>
    </main>
  );
}
```

## Using Material UI with a custom theme

### Theme Registry

> To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider` and the `useServerInsertedHTML` hook from `next/navigation` as follows:

Reviewer: Oxford comma 😇

```suggestion
To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider`, and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```

        inserted.push(serialized.name);
      }
      return prevInsert(...args);
    };
    const flush = () => {
      const prevInserted = inserted;
      inserted = [];
      return prevInserted;
    };
    return { cache, flush };
  });

  useServerInsertedHTML(() => {
    const names = flush();
    if (names.length === 0) {
      return null;
    }
    let styles = '';
    for (const name of names) {
      styles += cache.inserted[name];
    }
    return (
      <style
        key={cache.key}
        data-emotion={`${cache.key} ${names.join(' ')}`}
        dangerouslySetInnerHTML={{
          __html: styles,
        }}
      />
    );
  });

  return (
    <CacheProvider value={cache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
}

// app/layout.js
export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>
        <ThemeRegistry options={{ key: 'mui' }}>{children}</ThemeRegistry>
      </body>
    </html>
  );
}
```

### CSS injection order

<!-- https://github.com/emotion-js/emotion/issues/3059 -->

> By default, Emotion inject Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles – for example if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.

Reviewer: ```suggestion
By default, Emotion injects Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles—for example, if you're customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.
```

Emotion provides the `prepend: true` option for `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.

Currently, `prepend` does not work reliably with the App Router, but you can work around it by wrapping Emotion styles in a CSS `@layer` with a modification to the snippet above:

```diff
 useServerInsertedHTML(() => {
   const names = flush();
   if (names.length === 0) {
     return null;
   }
   let styles = '';
   for (const name of names) {
     styles += cache.inserted[name];
   }
   return (
     <style
       key={cache.key}
       data-emotion={`${cache.key} ${names.join(' ')}`}
       dangerouslySetInnerHTML={{
-        __html: styles,
+        __html: options.prepend ? `@layer emotion {${styles}}` : styles,
       }}
     />
   );
 });
```

## Function props

> Props passed from server components - for example `page.js` or other routing files - must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).

Reviewer: ```suggestion
Props passed from Server Components—for example, `page.js` or other routing files—must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

This works without any additional directives:

```jsx
// app/page.tsx
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';

export default function Page() {
  return (
    <>
      <Container maxWidth="lg">
        <Box>
          <Card raised>
            <Typography variant="h2">Hello World</Typography>
          </Card>
        </Box>
      </Container>
    </>
  );
}
```

> However, function props such as event handlers or render props are **non-serializable** - this won't work in a server component:

Reviewer: same as Joy comments for this section
